### PR TITLE
DC-138 - Socure DocV integration

### PR DESCRIPTION
**Pull request by necampanini**
# PR Review Priority Guide

- **PR:** #156 (DC-138 - Socure DocV integration)
- **Branch:** `feat/DC-138-socure-docv-integration` -> `main`
- **Risk Level:** High
- **Files:** 144 (55 code + 88 locale JSON deletions + 2 CSVs - 1 gitignore)
- **Lines:** +2697 / -2614
- **Generated:** 2026-04-13 12:32 EDT

Full Socure Document Verification (DocV) integration: new device-intelligence SDK on the frontend, PII-heavy ID-proofing submission with 3-attempt cap, an externally-callable webhook endpoint, expanded CSP, and a DB migration. Most of the line count is noise: 88 stale locale JSON deletions under `packages/design-system/content/locales/**` (gitignore cleanup) and two CSV content rewrites — skim those or skip. Real reviewer attention belongs on the Socure client, the `SubmitIdProofingCommandHandler`, the webhook endpoint + its rate-limit policy, and the CSP/camera-permission relaxation in `proxy.ts`.

---

## Priority Review Table

| Priority | File | Function/Change | Sec | Cpx | Nov | Notes |
|:---:|---|---|:---:|:---:|:---:|---|
| :red_circle: | `src/SEBT.Portal.UseCases/IdProofing/SubmitIdProofing/SubmitIdProofingCommandHandler.cs` | `Handle()` rewrite (+145) | High | High | Med | PII + attempt cap + race |
| :red_circle: | `src/SEBT.Portal.Api/Controllers/IdProofing/WebhookController.cs` | Anonymous webhook endpoint (+2 on controller; full surface in policy) | High | Med | High | External attack surface |
| :red_circle: | `src/SEBT.Portal.Infrastructure/Services/HttpSocureClient.cs` | `RunIdProofingAssessmentAsync` (+51) | High | Med | Med | PII in transit to Socure |
| :red_circle: | `src/SEBT.Portal.Web/src/proxy.ts` | CSP + Permissions-Policy (+17) | High | Low | High | Camera + Socure frame/script hosts |
| :red_circle: | `src/SEBT.Portal.Api/Program.cs` | Webhook rate-limit + DI wiring (+28) | High | Med | Med | New middleware path |
| :orange_circle: | `src/SEBT.Portal.Api/RateLimitPolicies.cs` + `WebhookRateLimitSettings.cs` (new) | Webhook policy constants | High | Low | Med | Threshold TBD |
| :orange_circle: | `src/SEBT.Portal.Web/src/features/auth/components/device-intelligence/di-adapter.ts` (new +83) | Dynamic script load from `sdk.dv.socure.io` | High | Med | High | 3rd-party script injection |
| :orange_circle: | `src/SEBT.Portal.Web/src/features/auth/components/device-intelligence/useDeviceIntelligence.ts` (new +35) | `useDeviceIntelligence` hook | Med | Low | High | Best-effort token retrieval |
| :orange_circle: | `src/SEBT.Portal.Web/src/features/auth/components/id-proofing/IdProofingWithDi.tsx` (new +23) | DI wrapper around `IdProofingForm` | Med | Low | High | First DI integration point |
| :orange_circle: | `src/SEBT.Portal.Web/src/features/auth/components/doc-verify/DocVerifyPage.tsx` (+63) | State-machine rewrite: stale-session detection, 15s capture timeout, webhook-safety-net polling | Med | High | Med | Concurrent state sources |
| :orange_circle: | `src/SEBT.Portal.Web/src/features/auth/components/doc-verify/VerificationPending.tsx` | 404-as-terminal, `SK_STILL_CHECKING` persistence | Med | Med | Low | sessionStorage semantics |
| :orange_circle: | `src/SEBT.Portal.Web/src/features/auth/components/doc-verify/sessionKeys.ts` (new) | `clearChallengeContext()` | Med | Low | Med | PII-adjacent session key mgmt |
| :orange_circle: | `src/SEBT.Portal.Core/AppSettings/SocureSettings.cs` (+15) | `DiSessionToken`, `SandboxPhoneOverride`, `WebhookSecret` fields | High | Low | Low | Secrets live here |
| :orange_circle: | `src/SEBT.Portal.Api/appsettings.json` + `appsettings.dc.example.json` (+16 / +4) | Webhook rate limit + Socure blocks | High | Low | Low | Verify prod `UseStub: false` |
| :orange_circle: | `src/SEBT.Portal.Infrastructure/Repositories/DatabaseDocVerificationChallengeRepository.cs` (+12) | `DbUpdateException` -> `DuplicateRecordException` on unique-index race | Low | Med | Med | Concurrency pattern |
| :orange_circle: | `src/SEBT.Portal.Core/Exceptions/DuplicateRecordException.cs` (new +15) | New exception type | Low | Low | Med | Escape-hatch across layers |
| :orange_circle: | `src/SEBT.Portal.Infrastructure/Migrations/20260409181556_AddIdProofingAttemptCountToUsers.cs` (new) | `+ IdProofingAttemptCount int default 0` | Med | Low | Low | Backfill safety on large tables |
| :orange_circle: | `test/SEBT.Portal.Tests/Unit/Infrastructure/Services/HttpSocureClientTests.cs` (new +405) | HttpClient disposal, PII field shaping | Low | Med | Med | Covers new surface; already patched for `using` (commit `b2f5b594`) |
| :orange_circle: | `test/SEBT.Portal.Tests/Unit/UseCases/IdProofing/SubmitIdProofingCommandHandlerTests.cs` (+377) | Max-attempts, race, household-lookup-failure, attempt counting | Low | Med | Low | Primary safety net for handler rewrite |
| :yellow_circle: | `src/SEBT.Portal.Web/src/features/auth/components/id-proofing/IdProofingForm.tsx` (+15) | `getDiToken` plumbing + `clearChallengeContext()` on submit | Med | Low | Low | Small |
| :yellow_circle: | `src/SEBT.Portal.Web/src/features/auth/components/doc-verify/adapters/socure-adapter.ts` (+18) | SDK launch args, ACCEPT persistence | Med | Low | Low | Small |
| :yellow_circle: | `src/SEBT.Portal.Web/src/features/auth/api/submit-id-proofing/schema.ts` (+12) | Optional `diSessionToken`, nullable response fields | Med | Low | Low | Zod boundary |
| :yellow_circle: | `src/SEBT.Portal.Web/src/features/auth/api/doc-verify/useVerificationStatus.ts` | Poll cap 30 -> 180 (~30 min window) | Low | Low | Low | Match challenge TTL |
| :yellow_circle: | `src/SEBT.Portal.Web/src/app/(public)/login/id-proofing/off-boarding/page.tsx` (+3) + test (+2) | New `challengeNotFound` reason wiring | Low | Low | Low | i18n + routing |
| :yellow_circle: | `src/SEBT.Portal.Web/src/env.ts` + `.env.example` (+2 / +5) | New DI SDK URL env + sandbox phone override | Med | Low | Low | Client-exposed config |
| :yellow_circle: | `packages/design-system/content/scripts/generate-locales.js` (+34 / -14) | SOURCE-column fallback | Low | Low | Med | See HEAD-only review-prep for detail |
| :yellow_circle: | `src/SEBT.Portal.Web/e2e/example.spec.ts` (+14 / -14) | CTA regex widening | Low | Low | Low | See HEAD-only review-prep |
| :green_circle: | `packages/design-system/content/locales/**` (88 files deleted) | Untracked generated output cleanup | Low | Low | Low | Skim |
| :green_circle: | `packages/design-system/content/states/{co,dc}.csv` (600/568 line rewrites) | Content-only; sheet-owned | Low | Low | Low | Content-owner review, not code review |
| :green_circle: | `src/SEBT.Portal.Infrastructure/Migrations/*Snapshot.cs` + `*.Designer.cs` | Auto-generated from migration | Low | Low | Low | Skim to confirm no manual edits |
| :green_circle: | Mock clients: `DisabledSocureClient.cs`, `StubSocureClient.cs` (+7 each) | Signature alignment with new params | Low | Low | Low | Skim |
| :green_circle: | `Core/Models/Auth/User.cs` (+6), `Data/Entities/UserEntity.cs` (+5), `Repositories/DatabaseUserRepository.cs` (+3), `Helpers/UserFactory.cs` (+1) | `IdProofingAttemptCount` propagation | Low | Low | Low | Mechanical |

---

## Summary by Priority

**High / Critical (5 items)** — Real external attack surface, PII flows, and platform-security config. Do not ship without reading each one end-to-end.
- `SubmitIdProofingCommandHandler.Handle()` — PII visibility mask, household-lookup-failure swallow, attempt-count increment, `DuplicateRecordException` race handling, retry-eligibility override. Highest cyclomatic complexity in the PR.
- `WebhookController` — externally-callable, `AllowAnonymous`-style. Confirm: (a) rate-limit policy actually applied via attribute/filter, (b) signature validation (`WebhookSecret`) is enforced on every non-stub path and uses a constant-time compare, (c) body-size cap + timeout.
- `HttpSocureClient.RunIdProofingAssessmentAsync` — every PII field the portal knows about is now sent to Socure; confirm no logging of raw request body, confirm `HttpClient` lifecycle (recent fix in `b2f5b594` added `using` in tests — verify production code path too), confirm failure modes don't expose inner exception messages upstream.
- `proxy.ts` CSP changes — widened `script-src` to `sdk.dv.socure.io`, `frame-src`/`child-src`/`style-src`/`font-src` to `verify-v2.socure.com`, `connect-src` to `*.socure.{com,io}`, and `Permissions-Policy: camera=(self "https://verify-v2.socure.com")`. Wildcards on `connect-src` are the one to argue about — everything else is specific hosts.
- `Program.cs` webhook wiring — new DI registration + middleware ordering. Verify rate-limit middleware precedes the controller and that the webhook route is not accidentally covered by an auth policy that would 401 Socure.

**Medium (14 items)** — Worth reading carefully but scoped.
- Device-intelligence SDK bootstrap + hook (`di-adapter.ts`, `useDeviceIntelligence.ts`, `IdProofingWithDi.tsx`) — first third-party script load pattern in the app. Confirm: SDK script URL is env-driven (not hardcoded), load failure is silent + non-blocking, no DI token cached in localStorage.
- `DocVerifyPage` state-machine rewrite — three concurrent state sources (URL param, sessionStorage, polled verification status). The 15s capture-to-pending timeout plus webhook-safety-net polling is the load-bearing fix for the "mobile capture never calls onSuccess" edge case; read alongside the new `DocVerifyPage.test.tsx` cases.
- `VerificationPending` — 404-as-terminal path and the `SK_STILL_CHECKING` sessionStorage flag; confirm no PII in session keys.
- `sessionKeys.ts` — `clearChallengeContext()` is called from both the form submit and the verify-page; confirm no keys are missed (session leak between attempts was the root cause of several fixes in recent commits).
- `SocureSettings` — `WebhookSecret`, `DiSessionToken`, `SandboxPhoneOverride`. All three are secrets. Confirm they are loaded from `/run/secrets` or env, never from committed appsettings, and that `SandboxPhoneOverride` is ignored outside dev.
- `appsettings.json` + `appsettings.dc.example.json` — verify example files contain no real secrets and that `UseStub: false` is explicit for non-dev.
- `DatabaseDocVerificationChallengeRepository` — confirms unique-index name matches entity config; confirm `DuplicateRecordException` is only thrown for the challenge index, not any `DbUpdateException`.
- New `DuplicateRecordException` type — confirm it doesn't leak DB-internal details when serialized through the exception middleware.
- Migration — `IdProofingAttemptCount int default 0` is safe for backfill, but verify EF's migration script emits a server-side default (not a client-side update), and that `Users` table doesn't have triggers that would fire on backfill.
- Backend tests — spot-check the max-attempts and race-handling cases in `SubmitIdProofingCommandHandlerTests.cs`; spot-check that `HttpSocureClientTests.cs` asserts PII fields are sent (i.e., coverage, not just that the call succeeds).
- `IdProofingForm.tsx` DI plumbing — DI token is added to the submit payload; confirm it's optional server-side and not required-for-success in the handler.
- Socure-adapter `socure-adapter.ts` — 18-line delta covering SDK launch args and ACCEPT outcome persistence; small but part of the critical DocV resume path.
- Submit-id-proofing Zod schema — nullable response fields are load-bearing for the new "best-effort household lookup" path; confirm every frontend consumer handles `null`.
- `useVerificationStatus` 30→180 poll cap — confirm the poll interval hasn't also changed, otherwise the effective window differs from intent.

**Low (10 items)** — Mostly mechanical or out-of-scope for deep review. Can be skimmed.
- 88 locale JSON deletions, 2 CSV rewrites (content-owner review, not code review).
- EF auto-generated snapshot + designer files.
- Stub/Disabled Socure clients — signature alignment only.
- `User`/`UserEntity`/`UserFactory`/`DatabaseUserRepository` attempt-count plumbing — mechanical.
- The two files from HEAD commit (locale generator + e2e regex) already have a focused review-prep in this directory.

---

## Review Recommendation

Work outside-in from the attack surface:

1. **Start at the webhook boundary.** Read `WebhookController.cs` + `Program.cs` + `RateLimitPolicies.cs` + `WebhookRateLimitSettings.cs` + the webhook block of `appsettings.json` together. Confirm: anonymous, rate-limited, signature-validated, body-size-capped. Settle this before touching handler internals — if the endpoint isn't locked down, nothing downstream matters.
2. **Then read `SubmitIdProofingCommandHandler.cs` top to bottom** with `SubmitIdProofingCommandHandlerTests.cs` side-by-side. Trace attempt cap, household-lookup swallow, and the `DuplicateRecordException` race path. Note whether every early return leaves user/challenge state consistent (attempt count bumped? challenge left Pending or cleaned?).
3. **Then read `HttpSocureClient.cs`** with its new tests. Verify HttpClient lifecycle, error-to-caller mapping, and that no PII lands in logs or upstream exception messages.
4. **Then jump to the frontend security surface:** `proxy.ts` CSP/permissions first (one-pass review, flag any wildcard you can't justify), then `di-adapter.ts` + `useDeviceIntelligence.ts` (confirm load-failure is silent and non-blocking). Flag any PII written to `localStorage` (repo rule: never).
5. **Then `DocVerifyPage.tsx` + `VerificationPending.tsx` + `sessionKeys.ts`** as a unit. The new tests in `DocVerifyPage.test.tsx` will tell you what edge cases the author considers load-bearing — reading those first is a fast orientation.
6. **Skim** the mechanical attempt-count plumbing, the migration snapshot, the stub client signature updates, and the locale-generator/e2e regex pair (already reviewed separately).

### CODE REVIEW

File: src/SEBT.Portal.Web/src/proxy.ts [hunk 1]
```diff
@@ -26,13 +26,13 @@ export function proxy(request: NextRequest) {
   // In dev, we skip style nonce to allow HMR style injection.
   const cspHeader = `
     default-src 'self';
-    script-src 'self' 'nonce-${nonce}' 'strict-dynamic' https://www.googletagmanager.com ${isDev ? "'unsafe-eval'" : ''};
-    style-src 'self' ${isDev ? "'unsafe-inline'" : `'nonce-${nonce}'`} https://fonts.googleapis.com;
-    font-src 'self' https://fonts.gstatic.com;
+    script-src 'self' 'nonce-${nonce}' 'strict-dynamic' https://www.googletagmanager.com https://sdk.dv.socure.io ${isDev ? "'unsafe-eval'" : ''};
+    style-src 'self' ${isDev ? "'unsafe-inline'" : `'nonce-${nonce}'`} https://fonts.googleapis.com https://verify-v2.socure.com;
+    font-src 'self' https://fonts.gstatic.com https://verify-v2.socure.com;
```
[the remark this answers is not in the document]
Comment (noxmwalsh) [reply src/SEBT.Portal.Web/src/proxy.ts:31]: This is actually accurate.  It will not load in dev (since it's 'production-like') either, so we should add that in

File: src/SEBT.Portal.Web/src/proxy.ts [hunk 2]
```diff
@@ -59,7 +59,10 @@ export function proxy(request: NextRequest) {
   response.headers.set('X-Frame-Options', 'DENY')
   response.headers.set('X-Content-Type-Options', 'nosniff')
   response.headers.set('Referrer-Policy', 'strict-origin-when-cross-origin')
-  response.headers.set('Permissions-Policy', 'camera=(), microphone=(), geolocation=()')
+  response.headers.set(
+    'Permissions-Policy',
+    'camera=(self "https://verify-v2.socure.com"), microphone=(), geolocation=()'
```
Comment (noxmwalsh) [inline src/SEBT.Portal.Web/src/proxy.ts:64]: Is this to support the mobile version?

File: src/SEBT.Portal.Web/src/features/auth/components/doc-verify/adapters/socure-adapter.ts
Comment (noxmwalsh) [inline src/SEBT.Portal.Web/src/features/auth/components/doc-verify/adapters/socure-adapter.ts:13]: I know this was here before, but would this need to be included in `proxy.ts` too?

File: src/SEBT.Portal.Web/src/env.ts [hunk 3]
```diff
@@ -32,6 +32,7 @@ export const env = createEnv({
     NEXT_PUBLIC_STATE: z.enum(['dc', 'co']),
     NEXT_PUBLIC_GA_ID: z.string().startsWith('G-').optional(),
     NEXT_PUBLIC_SOCURE_SDK_KEY: z.string().min(1).optional(),
+    NEXT_PUBLIC_SOCURE_DI_SDK_KEY: z.string().min(1).optional(),
```
Comment (noxmwalsh) [inline src/SEBT.Portal.Web/src/env.ts:35]: Doesn't this replace `NEXT_PUBLIC_SOCURE_SDK_KEY`?  I wasn't aware there was a separate Digital Intelligence Socure SDK key (at least from what I saw in the RiskOS interface)

**Comment by noxmwalsh** [comment]
Verified locally! Thanks for your patience while I was reviewing this Nick!

**Review by noxmwalsh — APPROVED** [review]
